### ISSUE TITLE
Fatal error: debugger does not support channel locks

### ISSUE THREAD

**Issue by anentropic** [ocaml/ocaml#10517]
I originally posted this on Stackoverflow, thinking it a problem with my code or some library incompatibility.
https://stackoverflow.com/questions/68394075/fatal-error-debugger-does-not-support-channel-locks

I had a reply there suggesting it was likely an issue with `ocamldebug` and macos and I should report it here, so I'll reproduce details below.

Firstly I am on macOS Catalina (10.15.7) on an Intel cpu.

I initially set up my project by:
```sh
brew install ocaml opam
opam switch create . ocaml-base-compiler.4.12.0
eval $(opam env)
```

I have added `(modes byte exe)` to my `dune` file.

When I run `dune build` I can see the bytecode file output, alongside the exe, as `_build/default/bin/cli.bc`

When I pass this to `ocamldebug` I get the following error:

```sh
ocamldebug _build/default/bin/cli.bc
	OCaml Debugger version 4.12.0

(ocd) r
Loading program... done.
Fatal error: debugger does not support channel locks
Lost connection with process 33035 (active process)
between time 170000 and time 180000
Restart from time 170000 and try to get closer of the problem ? (y or n)
```

If I choose `y` the console seems to hang indefinitely.

I found the source of the error here:  
https://github.com/ocaml/ocaml/blob/f68acd1a618ac54790a8347fad466084f15a9a9e/runtime/debugger.c#L144

```c
  /* The code in this file does not bracket channel I/O operations with
     Lock and Unlock, so fail if those are not no-ops. */
  if (caml_channel_mutex_lock != NULL ||
      caml_channel_mutex_unlock != NULL ||
      caml_channel_mutex_unlock_exn != NULL)
    caml_fatal_error("debugger does not support channel locks");
```

...but I don't know what might be triggering it.

My project is using `cmdliner` and `lwt` ...I thought "channel I/O operations" might have something to do with either of those libraries.

But the replier on SO seemed to think it was more likely an OS-level issue with my install of `ocamldebug` itself.

**Comment by nojb** [comment]
I believe the message `debugger does not support channel locks` refers to the fact that the debugger does not support programs that enable multi threading via the `threads` library. Perhaps `lwt` does this indirectly?

**Comment by dra27** [comment]
Does your project use `lwt.unix`?

**Comment by anentropic** [comment]
@dra27 I think so, yes... indirectly via https://github.com/mirage/irmin/

**Comment by xavierleroy** [comment]
We should have a better error message! But, yes, ocamldebug doesn't work with multithreaded programs.

**Comment by anentropic** [comment]
ok thanks, good to know

looks like I need to temporarily vendor `irmin` into my project and put in some print debugging then

**Comment by dra27** [comment]
If you are _certain_ (and I mean really, really certain) that no threads are ever created, for debugging purposes only, you could recompile `systhreads.cma` with:

https://github.com/ocaml/ocaml/blob/f203a5d45dd32216596d3de1f5fd02cd8cab6f7b/otherlibs/systhreads/thread.ml#L69

commented out or, possibly less unsafely, these 3 lines commented out:

https://github.com/ocaml/ocaml/blob/f203a5d45dd32216596d3de1f5fd02cd8cab6f7b/otherlibs/systhreads/st_stubs.c#L469-L471

commented out. But you must be _absolutely_ certain that no threads will be created.

**Comment by anentropic** [comment]
seems pretty clear that `lwt` uses threads https://github.com/ocsigen/lwt/search?q=threads so I guess i'm out of luck

**Comment by xavierleroy** [comment]
It would be nice if we could start ocamldebug on a program that was linked with the threads library and fail only when the first thread is actually created.

**Comment by anentropic** [comment]
Out of curiosity, is this situation going to get better or worse with multi-core OCaml?

**Comment by xavierleroy** [comment]
> It would be nice if we could start ocamldebug on a program that was linked with the threads library and fail only when the first thread is actually created.

See #10594

**Comment by xavierleroy** [comment]
> Out of curiosity, is this situation going to get better or worse with multi-core OCaml?

As far as I know, no better and no worse, i.e. the debugger should remain compatible with single-domain, single-thread programs only.

**Comment by xavierleroy** [comment]
#10594 is merged and adds a clearer error message + the ability to debug the program until it calls `Thread.create`.  I think this is the best we can do without an extensive rewrite of ocamldebug.  Hence I'll close this report.

**Comment by nikswamy** [comment]
I'm using OCaml 4.12. My program does have some uses of Thread.create/join, and Mutex.lock/unlock.

When running my program using ocamldebug on Linux, it fails very quickly with `Fatal error: debugger does not support channel locks`. 

However, running the same program on Windows/Cygwin does not exhibit this error and ocamldebug goes much further, until it eventually dies with `Fatal error: exception Sys_error("An existing connection was forcibly closed by the remote host.\r\n")`

Is there a reason for this difference in behavior? Thanks in advance!

**Comment by xavierleroy** [comment]
The Windows/Cygwin behavior you describe is consistent with OCaml 4.11 or earlier.  The "debugger does not support channel lock" error was introduced in 4.12 but should show up with Cygwin as well as with other systems.

The soon-to-be-released OCaml 4.14.0 should restore a 4.11-like behavior, namely: you can debug up to the point where the program creates its first thread.
